Fix portal surface blending, entity selection and recursive mirrors

VReaperV opened the pull request:
Fixes portal blending with portal surfaces, misc_portal_surface entity selection and GL_FrontFace for mirrors in chains of recursive portals/mirrors. Fixes #1011. Disables the 64qu distance limit for portal entities which wasn't working properly anyway.

Portal entities are now chosen based on their distance to the center of the portals surface (calculated as the average of all vertices). Old code was checking if they are within 64qu of the plane, which was resulting in portals entities that are actually much further away being chosen.

Introduces Tess_StageIteratorPortal(), intended for portals to properly write to stencil buffer.

Screenshots (some blended surfaces look slightly off, this might be an issue with lighting in general, portals themselves blend correctly):
![unvanquished_2024-01-09_210516_000](https://github.com/DaemonEngine/Daemon/assets/10687142/d1bf167c-d000-4ff4-a4b3-eecf6d9a88b2)
![unvanquished_2024-01-09_203209_000](https://github.com/DaemonEngine/Daemon/assets/10687142/06343710-39dd-41e7-96d1-b2e6da3c89ce)
![unvanquished_2024-01-09_195110_000](https://github.com/DaemonEngine/Daemon/assets/10687142/c6581233-9a38-48c2-9731-b81487eec1cd)
![unvanquished_2024-01-09_195048_000](https://github.com/DaemonEngine/Daemon/assets/10687142/1e33c1ab-f14b-420e-a7b2-2f8c8bd20705)
![unvanquished_2024-01-07_020443_000](https://github.com/DaemonEngine/Daemon/assets/10687142/9f51bb5e-c244-4bb2-8319-76d5560e04b7)
![unvanquished_2024-01-07_020416_000](https://github.com/DaemonEngine/Daemon/assets/10687142/003a1153-656d-4f61-941b-96410cf13a5c)
![unvanquished_2024-01-07_020340_000](https://github.com/DaemonEngine/Daemon/assets/10687142/ba2e0d0a-749e-4451-91c5-71c976eb3e45)

slipher commented:
Could you give an overview of how portal rendering works? Stuff like, does it render camera view to a separate buffer, or does it do some projection to draw them directly into the current scene?

slipher commented:
> Portal entities are now chosen based on their distance to the center of the portals surface (calculated as the average of all vertices). Old code was checking if they are within 64qu of the plane, which was resulting in portals entities that are actually much further away being chosen.

If you change how a portal surface's view location, does that mean you would need to update q3map2 as well? I believe it needs to understand this to calculate the PVS.

VReaperV commented:
> Could you give an overview of how portal rendering works? Stuff like, does it render camera view to a separate buffer, or does it do some projection to draw them directly into the current scene?

Sure. It does in fact just change projection.
1) In each R_SortDrawSurfs, all portal surfaces are rendered first after a depth pass (all surfaces with SS_PORTAL shader).
2) For each portal surface, a new viewParms is created that should represent the correct camera view. All render entities are iterated and the closest entity with type RT_PORTALSURFACE is chosen (old code was choosing the first such entity that was within 64qu of the plane the surface was on). The pvsOrigin and orientation of the new view is calculated, which is currently also done incorrectly (there will be a pr later to fix that).
3) Portal surface is drawn into depth and stencil buffer, stencil is increased by one, depth is set to max.
4) Portal frustums and projection matrix are set up to render the correct portal view onto the portal surface. The view is rendered which can cause another portal to render as above.
5) The portal surface is rendered again (back with the previous viewParms), first into depth and colour buffer to blend the portal view with the portal surface and to set the proper depth, then into stencil to decrease it by 1.
6) The rest of the view is rendered normally.

VReaperV commented:
> If you change how a portal surface's view location is selected, does that mean you would need to update q3map2 as well? I believe it needs to understand this to calculate the PVS.

No, since it uses the same code as any "normal" view rendering does, any viewpos on the map that has the correct pvs if you just go there or noclip to it will also have the correct pvs when rendering a portal. Even if it's e. g. a room that is "off-map" (not supposed to be able to get there or see it normally during gameplay), it still renders only what's in that room.

slipher commented:
> > If you change how a portal surface's view location is selected, does that mean you would need to update q3map2 as well? I believe it needs to understand this to calculate the PVS.
> 
> No, since it uses the same code as any "normal" view rendering does, any viewpos on the map that has the correct pvs if you just go there or noclip to it will also have the correct pvs when rendering a portal. Even if it's e. g. a room that is "off-map" (not supposed to be able to get there or see it normally during gameplay), it still renders only what's in that room.

The PVS is not (primarily) for the renderer. It's for the server to know which entities it needs to send to a client because the client might be able to see them. So the map compiler needs to know which areas can be seen through a portal, so it can add those areas to the PVS of the area with the portal texture.

slipher commented on src/engine/renderer/tr_backend.cpp:
```diff
@@ -5604,17 +5604,31 @@ const RenderCommand *FinalisePortalCommand::ExecuteSelf( ) const
+	rb_surfaceTable[Util::ordinal( *( surface->surface ) )]( surface->surface );
+	Tess_End();
+
 	glStencilFunc( GL_EQUAL, backEnd.viewParms.portalLevel + 1, 0xff );
```
The stencil calls now appear twice. Is that intentional?

VReaperV replied:
You mean the glStencilFunc() call? One of those can be removed, good call

VReaperV replied:
Assuming the above, fixed

slipher commented on src/engine/renderer/tr_main.cpp:
```diff
@@ -1301,40 +1301,48 @@ static bool R_GetPortalOrientations( drawSurf_t *drawSurf, orientation_t *surfac
 		// if the entity is just a mirror, don't use as a camera point
-		if ( e->e.oldorigin[ 0 ] == e->e.origin[ 0 ] && e->e.oldorigin[ 1 ] == e->e.origin[ 1 ] && e->e.oldorigin[ 2 ] == e->e.origin[ 2 ] )
-		{
+		if ( e->e.oldorigin[0] == e->e.origin[0] && e->e.oldorigin[1] == e->e.origin[1] && e->e.oldorigin[2] == e->e.origin[2] ) {
```
It would be preferable to avoid whitespace changes on otherwise unmodified lines

VReaperV replied:
Fixed.

slipher commented on src/engine/renderer/tr_shade.cpp:
```diff
@@ -2882,6 +2882,89 @@ void Tess_StageIteratorGeneric()
+			case stageType_t::ST_PORTALMAP:
+			case stageType_t::ST_HEATHAZEMAP:
+			case stageType_t::ST_LIQUIDMAP:
+				Render_generic( pStage );
```
What's the motivation of creating a new StageIterator function and why do we want to use Render_generic on things which are usually rendered with a different function?

VReaperV replied:
When writing stencil the portal surface should only be rendered once, otherwise it will be rendered with light stages and the like, which would break stencil. (https://github.com/DaemonEngine/Daemon/pull/1031/commits/d0ecf605d3f69f03dac57ef3acce2e1d4a745f19 updates it slightly to make sure only one stage is rendered)
Render_generic seemed to be the most appropriate there since only depth and stencil output are required at that stage, no extra information.

VReaperV replied:
The depth only pass can be changed to Tess_StageIteratorDepthFill() though.

VReaperV replied:
Actually, even that shouldn't be changed since otherwise it might do a wasteful ST_COLLAPSE_lighting_PBR stage render.

VReaperV commented:
> The PVS is not (primarily) for the renderer. It's for the server to know which entities it needs to send to a client because the client might be able to see them. So the map compiler needs to know which areas can be seen through a portal, so it can add those areas to the PVS of the area with the portal texture.

It seems to get the correct entities from my testing.

VReaperV commented:
Ah, that's good!

slipher commented on src/engine/renderer/tr_shade.cpp:
```diff
@@ -2882,6 +2882,89 @@ void Tess_StageIteratorGeneric()
+			break;
+		}
+
+		if ( r_showLightMaps->integer && pStage->type == stageType_t::ST_LIGHTMAP ) {
```
This can be removed since it already always breaks out of the loop

VReaperV replied:
Fixed.

illwieckz commented:
In the chat you said there would also be a patch on game code side, so I don't know if the bugs I see when running this branch are happening only because I miss that other patch, but on my end some portals are acting as mirrors.

Exemple with the the mirror-mirror-portal-mirror chain that is now a mirror-mirror-mirror chain:

map: `test-portal-recursion`
viewpos: `1397 -315 16 0 0`

![unvanquished_2024-01-18_152516_000](https://github.com/DaemonEngine/Daemon/assets/2311649/28777019-24af-401c-a10f-627b1fb3a3b6)

![unvanquished_2024-01-18_153723_000](https://github.com/DaemonEngine/Daemon/assets/2311649/8fb7850a-44fb-44fb-ac6d-cf526a58dadd)

Example in the portal+mirror room, the one on the left (the one in front of you when entering the room) should be a portal displaying the camera from the room that is next to it on the right:

map: `test-portal-recursion`
viewpos: `-1287 464 74 45 0`

![unvanquished_2024-01-18_153028_000](https://github.com/DaemonEngine/Daemon/assets/2311649/a113c533-fa7b-4c99-930e-ba2c552211e8)

The room next to it on the left of this one is a mirror+mirror room and it's expected to behave like this, but not this one.

illwieckz commented:
As a remind of what is a mirror and what is a portal, the portal entities without camera links are mirrors:

![20240118-153953-000 test-portal-recursion](https://github.com/DaemonEngine/Daemon/assets/2311649/25431c99-c650-403a-b375-48b0f14151ff)

illwieckz commented:
Also I spotted there a missing wall in mirror, in alien base, viewpos `-3010 -65 60 -60 0`:

![unvanquished_2024-01-18_155228_000](https://github.com/DaemonEngine/Daemon/assets/2311649/dc5e520d-d38a-4c3b-be02-f46baddcd2d8)

VReaperV commented:
> In the chat you said there would also be a patch on game code side, so I don't know if the bugs I see when running this branch are happening only because I miss that other patch, but on my end some portals are acting as mirrors.

Thank you, I removed the portalCenter initialization when I was reverting the other changes to make this pr since I never saw them get initialized anywhere else and thought they would be initialized to {0.0, 0.0, 0.0}. Fixed that now.

Broken mirror on "1397 -315 16 0 0" also fixed, as well as the room with a portal and a mirror.

slipher commented on src/engine/renderer/tr_shade.cpp:
```diff
@@ -2882,6 +2882,74 @@ void Tess_StageIteratorGeneric()
+			case stageType_t::ST_HEATHAZEMAP:
+			case stageType_t::ST_LIQUIDMAP:
+				Render_generic( pStage );
+				stageFound = true;
```
`return` ?

VReaperV replied:
Thanks, fixed.

slipher reviewed (commented):
LGTM

illwieckz commented:
The blending issue in maps like school is because the surface is not set as `bspSurface` and has `lightmapNum` sets to `-1` (no lightmap), this can be adressed in another PR.

illwieckz reviewed (approved):
This PR looks ready to me too as a “big first step”, remaining problems can be adressed in subsequent PRs.

sweet235 commented:
> This PR looks ready to me too as a “big first step”, remaining problems can be adressed in subsequent PRs.

I agree. We should not get stuck in trying to improve one PR until the end of days.

illwieckz commented:
Once CI has verified everything, you can modify the merge option and select “Rebase and merge”, this will avoid the creation of an useless merge commit.

VReaperV commented:
Allright, will do.
